content/tutorial/02-advanced-orders/01-eth-flow/03-cancel-eth-flow/README.md

```diff
@@ -2,4 +2,179 @@
 title: Cancelling order
 ---
 
-Something
\ No newline at end of file
+In the previous tutorials we have created an Eth-flow order and showed how to view it. Let's say for some reason we want to cancel the order (like instead of buying 1 xDAI worth of COW, you now want to buy 200 xDAI worth of COW)! In this tutorial we will show how to cancel an order.
+
+In order to cancel an Eth-flow order, we need to pass to `invalidateOrder` the original parameter that was passed to `createOrder`. There are a couple of ways to do this:
+
+- Lookup the `calldata` of the transaction that created the order, and replace the function selector with the `invalidateOrder` function selector
+- Lookup the logs of the transaction that created the order and reconstruct the `EthFlowOrder.Data` struct from the `OrderPlacement` event log
+
+We will use the second method in this tutorial.
+
+## Reconstructing the `EthFlowOrder.Data` struct
+
+From the [`EthFlowOrder.Data` struct](https://beta.docs.cow.fi/cow-protocol/reference/contracts/periphery/eth-flow#ethfloworderdata) technical reference, we can see that the struct's fields are a subset of the [`GPv2Order.Data` struct](https://beta.docs.cow.fi/cow-protocol/reference/contracts/core/settlement#gpv2orderdata-struct) and an additional `int64` field called `quoteId`.
+
+Fortunately, in the `OrderPlacement` event log, we can see that the `order` field is the `GPv2Order.Data` struct, and the `data` field is a solidity-packed tuple containing the `quoteId` field and the `validTo` field. This means that we can reconstruct the `EthFlowOrder.Data` struct from the `OrderPlacement` event log!
+
+### Extract `quoteId` from `data` field
+
+First let's write a helper function to extract the `quoteId` from the solidity packed `data` log field:
+
+```typescript
+/// file: run.ts
+import type { Web3Provider } from '@ethersproject/providers'
++++import { BigNumber, Contract, utils } from "ethers";+++
+import { SupportedChainId, UnsignedOrder } from '@cowprotocol/cow-sdk'
+import { onchainOrderToHash } from '/src/lib/gpv2Order';
+import abi from './ethFlow.abi.json'
+
+// ...
+
+export async function run(provider: Web3Provider): Promise<unknown> {
+    // ...
+
+    const decodePackedData = (data: string) => {
+        // Ensure data is of correct length (8 bytes for int64 + 4 bytes for uint32)
+        if (data.length !== 2 * (8 + 4) + 2) { // +2 for '0x'
+            throw new Error('Invalid data length');
+        }
+
+        // Extract quoteId (int64) and validTo (uint32) from the data
+        const quoteIdBytes = data.slice(2, 2 + 16); // 8 bytes for int64
+        const validToBytes = data.slice(2 + 16); // 4 bytes for uint32
```

Comment:
Is the math correct in these `slice`s?

`quoteId` will contain the chars from position 2 to 19
`validTo` will contain the chars from position 0 to 17

Keep in mind [`slice` returns a copy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice)

Reply:
`validTo` should contain chars from position 18 onwards, yes? (it's only a single argument being passed, ie. the `start`), therefore from zero-index position 18 (inclusive) until the end of the string.

Reply:
Ahhhhhhhhhhhhhhhhh 🤦 

Let's pretend that didn't happen...

Starting over....

----

Hey, giving that `quoteId` goes from 2 to 18, shouldnt `validTo` start after that?
I'm rarely wrong as the track record states, but looks like it's reusing the last 2 chars from `quoteId`, since the `0x` chars are not discounted:
[0x][quoteId][validTo]
 2   + 2+16   + ....
 
 So here we should start from 
```suggestion
        const quoteIdBytes = data.slice(2, 2 + 16); // 8 bytes for int64
        const validToBytes = data.slice(2 + 2 + 16); // 4 bytes for uint32
```

No?

Reply:
For posterity: I was being doubly dumb and not really understanding the `slice` fn api.
All is well.